java/ql/src/experimental/Security/CWE/CWE-089/MyBatisMapperXmlSqlInjectionLib.qll

```diff
@@ -0,0 +1,140 @@
+/**
+ * Provide classes for SQL injection detection in MyBatis Mapper XML.
+ */
+
+import java
+import MyBatisCommonLib
+import semmle.code.xml.MyBatisMapperXML
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.frameworks.Properties
+
+/** A sink for MyBatis Mapper method call an argument. */
+class MyBatisMapperMethodCallAnArgument extends DataFlow::Node {
+  MyBatisMapperMethodCallAnArgument() {
+    exists(MyBatisMapperSqlOperation mbmxe, MethodAccess ma |
+      mbmxe.getMapperMethod() = ma.getMethod()
+    |
+      ma.getAnArgument() = this.asExpr()
+    )
+  }
+}
+
+/** Get the #{...} or ${...} parameters in the Mybatis mapper xml file */
+private string getAnMybatiXmlSetValue(XMLElement xmle) {
+  result = xmle.getTextValue().trim().regexpFind("(#|\\$)\\{[^\\}]*\\}", _, _)
+}
+
+predicate isMapperXmlSqlInjection(DataFlow::Node node, XMLElement xmle) {
+  // MyBatis Mapper method Param Annotation sql injection vulnerabilities.
+  // e.g. MyBatis Mapper method: `void test(@Param("orderby") String name);` and MyBatis Mapper XML file:`select id,name from test order by ${orderby,jdbcType=VARCHAR}`
+  exists(MyBatisMapperSqlOperation mbmxe, MethodAccess ma, int i, Annotation annotation |
```

Comment:
Same question as for the annotation version -- what syntax are you looking for in each case, and why is each case necessary?

Reply:
I refactored this part. Take a look at the refactored.
https://github.com/github/codeql/pull/6319/commits/6c6113b85b55af120c259210ef1f24ca9d97b2e8#diff-51a2d0c14bf4df6c8350168ff95dd4e7fb6ceb2dfd166609cab36f8ed43f5240R81-R178